### PR TITLE
Script bindings for sockets

### PR DESCRIPTION
**Pull request by ahigerd**
This PR provides an event-driven TCP socket API for mGBA scripting. Where the behaviors overlap, it roughly imitates the LuaSocket API, but due to the mGBA scripting core not being yieldable, it can't be a perfect match.

Also included are two sample scripts. Both scripts will log messages received over the socket to the console and will send an input viewer. One script will connect to a server; the other will open a server and listen for incoming connections.

Documentation still needs to be written.

### CODE REVIEW

File: include/mgba-util/socket.h
Comment (endrift) [inline include/mgba-util/socket.h:447]: Perhaps make this a #warning, especially in the #else case.

File: include/mgba-util/socket.h
Comment (endrift) [inline include/mgba-util/socket.h:455]: Since `error` is the end of the functionin this case, an early return is preferred.

File: src/core/scripting.c
Comment (endrift) [inline src/core/scripting.c:821]: Consider splitting this off and also the whitespace fixes into a separate PR, especially since those are the only changes for this file.

File: src/script/engines/lua.c
Comment (endrift) [inline src/script/engines/lua.c:272]: Don't use printf. I should check if there's an mLOG category for scripting and add it if not.

File: src/script/socket.c
Comment (endrift) [inline src/script/socket.c:39]: I believe C99 lets you do `*ptr = (struct mScriptSocket) { ... }`

File: src/script/socket.c
Comment (endrift) [inline src/script/socket.c:153]: `const`

File: src/script/socket.c
Comment (endrift) [inline src/script/socket.c:179]: `const`

**Review by endrift — COMMENTED** [review]
Just some nits for now. Will take a closer look later, hopefully today.

File: src/script/engines/lua.c
Comment (ahigerd) [reply src/script/engines/lua.c:272]: Hmm... You know, I was just intending to remove that. It was just there for debugging. But if you think mLOG is a better idea, then sure.

File: src/script/socket.c
Comment (ahigerd) [reply src/script/socket.c:39]: I was just imitating how you'd written it elsewhere.

File: include/mgba-util/socket.h
Comment (ahigerd) [reply include/mgba-util/socket.h:455]: It's not the end of the function. It calls `freeaddrinfo` first.

Comment (ahigerd) [reply include/mgba-util/socket.h:455]: Oh wait. Wrong ifdef branch.

Huh. I thought I removed that one already. Must have rolled back.

File: src/script/engines/lua.c
Comment (ahigerd) [reply src/script/engines/lua.c:272]: _mLOG_CAT_SCRIPT btw

File: include/mgba-util/socket.h
Comment (ahigerd) [reply include/mgba-util/socket.h:447]: Why would there be a warning in the #else case?

Comment (endrift) [reply include/mgba-util/socket.h]: To denote that that case is strongly recommended against. It shouldn't be hit in any existing compilation path atm, so it'd be a good idea to tell people _don't hit this path_ if at all possible.

File: src/script/socket.c
Comment (endrift) [reply src/script/socket.c:39]: Don't worry about it then. I can clean up minor tech debt later.

**Comment by endrift** [comment]
Seems to be decent so far, but I'm curious what about it doesn't match, and uh, how the doc generator would handle it. I'm not really sure what the surface API would look like, to be honest. An example script is probably a good idea,

**Comment by ahigerd** [comment]
LuaSocket's `receive` API also has some oddly specific behaviors: https://w3.impa.br/~diego/software/luasocket/tcp.html#receive -- `*l` is useful for line-based protocols, sure, but `*a` is asking for a headache thanks to blocking. It also supports a second parameter that's just prefixed onto the return value. I chose not to implement any of these.

Also, LuaSocket isn't event-driven, so there's no callback API at all.

As far as the doc generator is concerned... Yeah, I have no idea.

I sent an example script in Discord, but I'll reproduce it here:
```lua
s = socket.tcp()
s:on("received", function()
  while true do
    local p, err = s:receive(1024)
    if p then
      console:log(p:match("^(.-)%s*$"))
    else
      if err ~= "temporary failure" then
        console:error(err)
      end
      return
    end
  end
end)
if s:connect("127.0.0.1", 8888) then
  console:log("Connected")
else
  console:log("Failed to connect")
end
```
It connects to a server and echoes anything it receives to the console.

**Comment by endrift** [comment]
I meant putting an example script _in the repo_. Also if you could build the doc generator and give me the yaml output I could see how it does.

File: src/script/socket.c
Comment (endrift) [inline src/script/socket.c:120]: Is this comment still accurate?

File: src/script/socket.c
Comment (endrift) [inline src/script/socket.c:175]: Should these be initial caps? I'm not used to seeing all lowercase error messages.

File: src/script/socket.c
Comment (endrift) [inline src/script/socket.c:145]: Should probably fill these in before merging.

File: src/script/socket.c
Comment (ahigerd) [reply src/script/socket.c:120]: Largely, yes. It would be better to collect all of the active sockets and select on them in one go instead of having one select() call per active socket. It's not the worst thing in the world because I wouldn't expect one of these scripts to have a lot of sockets open, of course.

File: src/script/socket.c
Comment (ahigerd) [reply src/script/socket.c:175]: It's convention from LuaSocket; they have to be exact strings because script code has to check for them.

File: src/script/socket.c
Comment (ahigerd) [reply src/script/socket.c:145]: True.

**Comment by ahigerd** [comment]
@endrift By the way, I noticed that resetting the scripting state leaves any open socket resources orphaned. We should probably figure something out for that...

Also we should discuss whether or not we want SO_REUSEADDR.

File: src/script/socket.c
Comment (ahigerd) [reply src/script/socket.c]: Updated it.

File: src/script/socket.c
Comment (endrift) [reply src/script/socket.c:199]: I wonder if the textual descriptions should be moved into the Lua implementation then, not the generic implementation that Lua sits on top of.

File: src/script/engines/lua.c
Comment (endrift) [inline src/script/engines/lua.c:335]: No definitions in `for` loops, though it's technically allowed in C99 I don't do it in mGBA code. I may consider changing this at some point though, especially now that I'm starting to move mGBA to C11.

File: src/script/socket.c
Comment (endrift) [inline src/script/socket.c:67]: Ditto.

File: include/mgba/internal/script/socket.h
Comment (ahigerd) [reply include/mgba/internal/script/socket.h:22]: I know that's the typical convention and that's what I did the first time, but then I discovered that no-`m` is the convention for namespaced constants exposed to the scripting engine.

Comment (ahigerd) [reply include/mgba/internal/script/socket.h:22]: Can change it to `SOCKERR` though.

File: src/script/socket.c
Comment (ahigerd) [reply src/script/socket.c:274]: Whoops. Good catch.

File: src/script/socket.c
Comment (ahigerd) [reply src/script/socket.c:69]: ...... I could have sworn I typed something there. *facepalm*

File: src/script/engines/lua.c
Comment (endrift) [inline src/script/engines/lua.c:8]: What uses `core/core.h`?

File: src/script/engines/lua.c
Comment (endrift) [inline src/script/engines/lua.c:22]: This is already done in `core/scripting.h`--though it should maybe be moved to `script/context.h` if you're using it here (there's a dependency graph between components, but I haven't actually explicitly defined it anywhere)

File: include/mgba/internal/script/socket.h
Comment (endrift) [inline include/mgba/internal/script/socket.h:22]: Though they're internal only so it probably doesn't matter that much, you can take a look at the PLATFORM exports to see how to export a constant that has a different name from its exported namespace.

**Review by endrift — COMMENTED** [review]
A few last comments before merging.

File: src/script/engines/lua.c
Comment (ahigerd) [reply src/script/engines/lua.c:8]: Nothing, anymore, because I decided to hook the frame callback in Lua code instead of in C code. I forgot about the include.

File: include/mgba/internal/script/socket.h
Comment (ahigerd) [reply include/mgba/internal/script/socket.h]: Done

File: src/script/engines/lua.c
Comment (ahigerd) [reply src/script/engines/lua.c:22]: Moved.

**Comment by ahigerd** [comment]
I switched the example scripts over to use `socket.ERRORS` instead of literal strings. As a quality-of-life tweak I also made the `socket.ERRORS` table respond to both numeric and string keys.

File: include/mgba/script/context.h
Comment (endrift) [inline include/mgba/script/context.h:11]: this goes after the CXX_GUARD

File: src/script/engines/lua.c
Comment (endrift) [reply src/script/engines/lua.c:22]: You should remove it here then.

File: src/script/socket.c
Comment (endrift) [inline src/script/socket.c:8]: If tihs isn't used you should remove it.

**Review by endrift — CHANGES_REQUESTED** [review]
Looks like some things only got half-addressed.

**Comment by ahigerd** [comment]
The heck? I did that. O.o Did I manage to miss committing something?

**Comment by endrift** [comment]
I'm ready to merge this unless you have any last-minute changes you want to make.

**Comment by ahigerd** [comment]
No complaints from me!

**Pull request merged**
